homeassistant/components/melnor/switch.py

```diff
@@ -21,52 +44,64 @@ async def async_setup_entry(
     async_add_devices: AddEntitiesCallback,
 ) -> None:
     """Set up the switch platform."""
-    switches = []
+    switches: list[MelnorZoneSwitch] = []
 
     coordinator: MelnorDataUpdateCoordinator = hass.data[DOMAIN][config_entry.entry_id]
 
     # This device may not have 4 valves total, but the library will only expose the right number of valves
     for i in range(1, 5):
         if coordinator.data[f"zone{i}"] is not None:
-            switches.append(MelnorSwitch(coordinator, i))
+
+            switches.append(
+                MelnorZoneSwitch(
+                    coordinator,
+                    i,
+                    MelnorSwitchEntityDescription(
+                        device_class=SwitchDeviceClass.SWITCH,
+                        icon="mdi:sprinkler",
+                        key="manual",
+                        name="Manual",
+                        on_fn=lambda valve: setattr(valve, "is_watering", True),
+                        off_fn=lambda valve: setattr(valve, "is_watering", False),
+                        state_fn=lambda valve: valve.is_watering,
+                    ),
+                )
+            )
 
     async_add_devices(switches)
 
 
-class MelnorSwitch(MelnorBluetoothBaseEntity, SwitchEntity):
+class MelnorZoneSwitch(MelnorZoneEntity, SwitchEntity):
     """A switch implementation for a melnor device."""
 
-    _attr_icon = "mdi:sprinkler"
+    entity_description: MelnorSwitchEntityDescription
 
     def __init__(
         self,
         coordinator: MelnorDataUpdateCoordinator,
-        valve_index: int,
+        zone_id: int,
+        entity_description: MelnorSwitchEntityDescription,
     ) -> None:
         """Initialize a switch for a melnor device."""
-        super().__init__(coordinator)
-        self._valve_index = valve_index
+        super().__init__(coordinator, zone_id)
 
-        valve_id = self._valve().id
-        self._attr_name = f"Zone {valve_id+1}"
-        self._attr_unique_id = f"{self._device.mac}-zone{valve_id}-manual"
+        self._attr_unique_id = f"{self._device.mac}-zone{zone_id}-manual"
+        self.entity_id = ENTITY_ID_FORMAT.format(self._attr_unique_id)
```

Comment:
I'm setting this manually because by default it wants to set it to something like `switch.zone_1` which is fairly generic, and if you have more than 1 of these devices (like me) you get entity id collisions on the second one.

So by including the parent device you get a unique and identifiable entity id right out of the gate.

Reply:
Setting the entity id manually is not usually done in new integrations.

https://developers.home-assistant.io/blog/2022/07/10/entity_naming

Reply:
Yeah, I'd seen that blog and I'm following those recommendations generally speaking. Is the intended behavior for nested device services complete? Consider the following device topology as currently implemented here (excluding the custom `entity_id`)

Right now if I have 2 of these devices (`FAKE_MAC_1` and `FAKE_MAC_2`) I'll have 2 device trees.

```
FAKE_MAC_1
│   Battery - sensor.fake_mac_1_battery
│
└───Zone 1
│   │   switch.zone_1_manual
|   |
└───Zone 2
    │   switch.zone_1_manual
FAKE_MAC_2
│   Battery - sensor.fake_mac_2_battery
│
└───Zone 1
│   │   switch.zone_1_manual_2
|   |
└───Zone 2
    │   switch.zone_1_manual_2
```

The sensors that are attached to the root device have a nice easily identifiable `entity_id` that maps you all the way back to the root device using the built-in naming behavior.

The zones, which follow the same behavior until they provide their own `DeviceInfo`, get an `entity_id` ignore the parent device entirely. If these were separate pieces of hardware that would make sense, but they're not. I could probably add the mac address to these child device names but that seems antithetical to the problem that the new world is trying to solve.

This is compounded in the way in which Melnor chose to handle device naming in that they didn't. Any user defined zone names are stored in-app rather than on-device, so I'll never have a less generic name than `Zone 1` available without breaking Home Assistant naming convention and including the device mac.

I understand the reasoning behind where Home Assistant is moving. I'm just trying to bridge a really awkwardly implemented piece of hardware with current Home Assistant best practices, even if that means implementing a few things in spirit rather than the letter of the law.

Reply:
For now I've tweaked the way I'm naming the zone devices.